lib/metadata_fields.rb

```diff
@@ -0,0 +1,215 @@
+		values = []
+		METADATA_FIELDS.keys.each { |k|
+			values << k.to_s
+		}
+		values << NAMESPACE_ALL
+	end
+
+	# output all fields regardless of namespace
+	def add_all_fields(multiple, fields)
+		METADATA_FIELDS.keys.each { |ns|
+			add_fields_by_ns(multiple, fields, ns)
+		}
+	end
+
+	# add fields specified by namespace
+	def add_fields_by_ns(multiple, fields, ns)
+		metadata_fields = METADATA_FIELDS[ns][:fields]
+		metadata_fields.each{ |fieldPair|
+			# if output is true then print out field name
+			# if multiple has the default value
+			# or one of the values in the yml file
+			output = ( multiple == DEFAULT_MULTIPLE ) ?
+					true :
+					metadata_fields[fieldPair][:multiple] == multiple
+
+			field = {}
+			field[:name] = fieldPair[0]
+			field[:attributes] = fieldPair[1]
+
+			fields.push(field) if output
+		}
+	end
+
+	def get_queryable_fields
+		fields = []
+		add_all_fields(DEFAULT_MULTIPLE, fields)
+
+		# Filter for queryable fields
+		fields = fields.select do |field|
+			if field[:attributes][:queryable].nil?
+				raise ArgumentError, "Queryable info not found for field \"#{field[:name]}\""
+			end
+
+			field[:attributes][:queryable] == true
+		end
+
+		fields
+	end
+
+	def get_facet_fields_in_display_order
+		get_fields_for_section_in_display_order(:facet)
+	end
+
+	def get_search_result_fields_in_display_order
+		get_fields_for_section_in_display_order(:search_result)
+	end
+
+	def get_detail_fields_in_display_order
+		get_fields_for_section_in_display_order(:detail)
+	end
+
+	def get_fields_for_section_in_display_order(section)
+		raise ArgumentError, '"section" argument is nil' if section.nil?
+
+		# Accept strings even though our data structure uses symbol keys
+		if section.is_a?(String)
+			section = section.to_sym
+		end
+
+		fields = []
+		add_all_fields(DEFAULT_MULTIPLE, fields)
+
+		# Filter for search_result fields
+		fields = fields.select do |field|
+			if field[:attributes][:display][section].nil?
+				raise ArgumentError, "section \"#{section}\" not found for field \"#{field[:name]}\""
+			end
+
+			field[:attributes][:display][section][:show] == true
+		end
+		# sort, then extract just display variables (flatten)
+		sort_fields_by_display_attribute(fields, section, :sort_key)
+		field_parts = get_field_display_variables(fields, section)
+
+		field_parts
+	end
+
+	def sort_fields_by_display_attribute(fields, section, sort_key)
+		fields.sort_by! do |field|
+			field[:attributes][:display][section][sort_key]
+		end
+
+		fields
+	end
+
+	def get_field_display_variables(fields, section)
+		# Gets the subset of variables relevant to controller/display for a section
+		field_parts = fields.map { |field|
+			{ :name => field[:name],
+				:solr_type => field[:attributes][:display][:solr_type],
+				:label => field[:attributes][:display][section][:label],
+				:special_handling => field[:attributes][:display][section][:special_handling]
+			}
+		}
+
+		field_parts
+	end
+
+	def get_solr_name_opts(field_part)
```

Comment:
Is `solr_type` always guaranteed to be either `"tesim"` or `"ssim"`?  If not, should assign default value to `solr_opts`